### PR TITLE
[ServiceNow] Add DLS for ServiceNow

### PR DESCRIPTION
**Pull request by parthpuri-elastic**
## Related to https://github.com/elastic/connectors/issues/1819

Add support of Document Level Security in ServiceNow.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

**Comment by moxarth-elastic** [comment]
buildkite test this

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/servicenow.py [hunk 1]
```diff
@@ -601,23 +724,83 @@ async def _fetch_table_data(self, batched_apis):
...
-            record_apis = self.servicenow_client.get_record_apis(
-                url=ENDPOINTS["TABLE"].format(table=service_name),
-                params={},
-                total_count=table_length,
+            async for batched_apis in self._get_batched_apis(service_name, params):
+                async for user in self._yield_table_data(batched_apis=batched_apis):
+                    yield user
+        except Exception as exception:
+            self._logger.warning(
+                f"Skipping table data for {service_name}. Exception: {exception}."
```
Comment (seanstory) [inline connectors/sources/servicenow.py:794]:
```suggestion
                f"Skipping table data for {service_name}. Exception: {exception}.", exc_info=True
```

File: connectors/sources/servicenow.py [hunk 2]
```diff
@@ -568,13 +667,29 @@ async def _attachment_metadata_producer(self, record_ids):
...
+            async for table_data in self.servicenow_client.get_data(
+                batched_apis=batched_apis
+            ):
+                for record in table_data:
+                    formatted_table_data = self._format_doc(data=record)
+                    serialized_table_data = self.serialize(doc=formatted_table_data)
+                    yield serialized_table_data
+        except Exception as exception:
+            self._logger.warning(
+                f"Skipping batch data for {batched_apis}. Exception: {exception}."
```
Comment (seanstory) [inline connectors/sources/servicenow.py:689]:
```suggestion
                f"Skipping batch data for {batched_apis}. Exception: {exception}.", exc_info=True
```

File: tests/sources/test_servicenow.py [hunk 3]
```diff
@@ -769,3 +797,116 @@ async def test_get_docs_with_advanced_rules(filtering):
...
+                            "sys_id": "id_1",
+                            "email": "admin@email.com",
+                            "user_name": "demo.user",
+                        }
+                    ]
+                ]
+            ),
+        ):
+            async for user in source.get_access_control():
+                assert user["_id"] == "id_1"
```
Comment (seanstory) [inline tests/sources/test_servicenow.py:824]: Can you make this a more robust assertion about the contents of `user`?

File: connectors/sources/servicenow.py [hunk 4]
```diff
@@ -51,13 +56,32 @@
...
+        "admin",
+        "sn_request_read",
+        "asset",
+        "atf_test_designer",
+        "atf_test_admin",
+    ],
+    "incident": ["admin", "sn_incident_read", "ml_report_user", "ml_admin", "itil"],
+    "kb_knowledge": ["admin", "knowledge", "knowledge_manager", "knowledge_admin"],
+    "change_request": ["admin", "sn_change_read", "itil"],
+}
```
[the remark this answers is not in the document]
Comment (seanstory) [reply connectors/sources/servicenow.py:71]: I'm noting that they're not the same as the mappings we had in Workplace Search:
```
      INCIDENTS = %w[admin incident_manager itil itil_admin major_incident_manager].freeze
      USERS = %w[admin user_admin rota_manager rota_admin].freeze
      KNOWLEDGE_ARTICLES = %w[admin knowledge knowledge_manager knowledge_admin].freeze
      REQUESTED_ITEMS = %w[admin itil].freeze
```

**Comment by seanstory** [comment]
Also please remember that this will need a corresponding Kibana PR to add the new RCF

**Comment by parthpuri-elastic** [comment]
buildkite test this

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/servicenow.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (parthpuri-elastic) [reply connectors/sources/servicenow.py:71]: The main reason is that I have added child roles, like users with `itil` roles have `sn_request_read`, `sn_change_read`, and `sn_incident_read` . So, users with `itil` have all these roles, but vice versa is not possible.

For the Users table, we are indexing all fields. But some fields are visible to admins only; So, I have added `admin` only.

File: connectors/sources/servicenow.py [hunk 5]
```diff
@@ -584,13 +699,21 @@ async def _fetch_table_data(self, batched_apis):
...
+                    table_data_with_access_control = self._decorate_with_access_control(
+                        document=serialized_table_data,
+                        access_control=table_access_control,
+                    )
                     await self.queue.put(
-                        (serialized_table_data, None)  # pyright: ignore
+                        (
+                            table_data_with_access_control,
+                            None,
+                        )  # pyright: ignore
```
[the remark this answers is not in the document]
Comment (parthpuri-elastic) [reply connectors/sources/servicenow.py]: Not needed, removed 👍

File: tests/sources/test_servicenow.py (hunk 3 above)
Comment (parthpuri-elastic) [reply tests/sources/test_servicenow.py]: yes, checked full response

**Comment by parthpuri-elastic** [comment]
> Also please remember that this will need a corresponding Kibana PR to add the new RCF

Yes, after this PR, I will raise PR in Kibana 👍

File: connectors/sources/servicenow.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhishekjoshi-crest) [reply connectors/sources/servicenow.py:71]: Actually, I have explored all the ACLs of these default services and found these roles. After finding all the roles, I have discarded the parent role, like `itil` and then we have tested all the services with the above roles.

[the remark this answers is not in the document]
Comment (seanstory) [reply connectors/sources/servicenow.py:71]: From Parthpuri in slack: 
> I have explored all the ACLs of these services and then found all the roles. Then discarded the parent roles like itil  and then we have come across with this mapping

**Review by seanstory — APPROVED** [review]

**Comment by seanstory** [comment]
buildkite test this

**Pull request merged**